### ISSUE TITLE
Severe issue involving latest Windows Insider builds

### ISSUE THREAD

**Issue by Alexei-Ilio** [TanninOne/modorganizer#1124]
For some reason not entirely clear to me and mentioned by others on various forums, there seems to be a major conflict with Mod Organizer and the latest versions of the Windows Insider program that completely blocks MO from working (i.e. ignoring mods used by profile being used and only using files directly in data folder). Though some have said it's due to these experimental builds implementing some more intense anti-malware measures that consider MO's virtual install system to be malware, disabling Windows Defender and other anti-virus programs does nothing, so I have no idea if this is really the source and I'm just inept on that topic (which is likely) or if it's something else. The reason I think this is worth reporting is that these builds include, as many have mentioned by now, a fix for the Windows 8/8.1 and 10 4064 mb memory cap for DX9 games. More important, however, is the fact that, if a fix is not found on either Windows side or MO's, this issue may carry on to the next commercial build and break MO entirely for Windows 10 users.

**Comment by feigi** [comment]
+1

**Comment by BlazingPL** [comment]
Do you know the lastest working build, and if its possible to revert windows to it? I'll try changing from fast ring to slow ring build and see if it works.

**Comment by feigi** [comment]
No the latest insider built no. But I installed the current insider built and rolled back to my non-insider afterwards.

**Comment by Alexei-Ilio** [comment]
The problem exists with both slow and fast ring. I just upgraded to Windows 10 so unfortunately I don't know the latest build it works on, but that doesn't really matter since you can't choose specific builds to update or roll back to.

**Comment by chrisbro-MSFT** [comment]
@TanninOne, I sent you a PM on the Nexus forums tonight asking you to send someone at Microsoft an email to try to see if we can find a solution for this; just crossposting here in hopes of reaching you.

**Comment by TanninOne** [comment]
Could someone please clarify if this is about MO1 or MO2?

**Comment by draxionangel** [comment]
@TanninOne Definitely affecting MO1 and Oldrim.
MO2 + SSE seem to work still (not 100% sure on this as I haven't actually personally tested this).

Timeline:
Build 16232 - 4GB VRAM Fix - MO1 + Oldrim working w/ 4GB fix
Build 16237 - No change - MO1 + Oldrim working w/ 4GB fix
Build 16241 - BROKEN - MO1 no longer loading mods into Oldrim
Build 16251 - Still broken
Build 16257 - Still broken
Build 16273 - Just came out yesterday, haven't updated yet, will try today

More threads/info:
https://www.reddit.com/r/skyrimmods/comments/6n87if/do_the_latest_windows_insiders_16241_build_cause/
https://www.reddit.com/r/skyrimmods/comments/6ti7ji/windows_insider_preview_build_to_be_released_in/
https://www.reddit.com/r/skyrimmods/comments/6n8pce/the_lastest_windows_10_inside_preview16241/

http://forum.step-project.com/topic/12353-just-posted-this-on-windows-10-insider-feedback-hub-please-go-upvote/

There's also a few issues on the Windows Insider Feedback Hub:
https://aka.ms/G3xdt7
https://aka.ms/w8hk7d
https://aka.ms/abo59r
https://aka.ms/Wwxlux

The only problems I could find relating to MO2 & the new Windows Insider builds are the following which are actually more about Wrye Bash so not sure if it's the same issue or not:
https://www.reddit.com/r/skyrimmods/comments/6n4lh4/help_mod_organizer_2_and_latest_windows_10/
http://forum.step-project.com/topic/12328-wryes-bash-wont-run-from-inside-mo2/
http://forum.step-project.com/topic/12162-windows-7-update-possibly-broke-the-usvfs-proxy/

**Comment by chrisbro-MSFT** [comment]
@draxionangel - Any chance that you could attach a native debugger to Mod Organizer and grab a dump?  From the other threads it looks possible that there's an AV happening somewhere in the code called by CreateFileW_rep.  I don't see symbols available for MO so you might have to build from source...

If the crash is in MO code then it can be fixed here.  If the crash is in Windows code we might be able to get it to a Windows dev who might be able to fix it in time for the Fall update or find a workaround in MO code.

**Comment by TanninOne** [comment]
I think I still have the pdbs so if someone could send me a hook.dll.dmp I should be able to debug it or I can share the pdb if you want to look yourself.

**Comment by TucoPacifico1980** [comment]
Keeping my fingers crossed on that one guys. The insider build brings more VRAM to DX9. And out of the sudden this... Once I soon join the insider program, I'll post you the dmp @TanninOne . Hope you receive it ASAP from someone already in insider.

**Comment by TanninOne** [comment]
Ok, I've identified the issue and it's entirely on the MO side of things, I wouldn't bet on MS reverting the change that broke it, though it may still be related to it being a preview release.

The good news is that it's a fixable problem in MO, should be doable in an hour or two. But as you know I can't actually work on MO any more so we'd need to find someone willing to be a hero and put the time in.

**Comment by Alexei-Ilio** [comment]
What's required in the way of knowledge of coding? I have a bit of experience with python and a minuscule amount in java, but if time is pretty much the only problem, I'd be willing to see if I can do it.

**Comment by draxionangel** [comment]
@chrisbro-MSFT Unfortunately I'm still pretty new to lower level coding/debugging and am not 100% sure how to go about this (I'm normally a web developer). I'm working on seeing if I can get MO to build from source as a first step but not sure how far I'll actually get with this.

@TanninOne Any chance you could go into more detail about the issue? Not that I would be able to fix it but I am fairly comfortable reading over c++ code and getting a gist of what's going on and I am just interested in hearing what the issue is.

**Comment by chrisbro-MSFT** [comment]
With help from Tannin and another MS developer we've tracked down the issue in MO.  In the current preview releases of Win10, the FindFirstFileExW hooks in MO's hook.dll aren't getting called.  Those methods were refactored in the Windows kernel.  The suggested fix would be to hook the other FindFirstFile* methods in addition to FindFirstFileExW.  That would probably mean writing other hook methods very similar to https://github.com/TanninOne/modorganizer-hookdll/blob/ff83aeffcbaa6f22216a9efd0190ba992b9f529a/src/dllmain.cpp#L575.

**Comment by Alexei-Ilio** [comment]
Welp, never mind, I couldn't even begin to comprehend how to fix that. Damn.

**Comment by TanninOne** [comment]
Exactly. So there are 4 FindFirstFile functions in the windows api: FindFirstFileA, FindFirstFileW, FindFirstFileExA and FindFirstFileExW.
Previously, the first 3 were implemented as "proxies" to FindFistFileExW, so they would convert their input parameters, call ...ExW then convert the output back to what they needed.
To save myself time and avoid bugs I only implemented a hook for ...ExW even though the gamebryo games actually call FindFistFileA because this way I caught all FindFirstFile requests and this worked in WinXP all the way to Windows 10, up to this latest release. Now the FindFirstFileA call no longer goes through FindFirstFileExW.

To fix it, you want to add a hook to FindFirstFileA and implement the hook the same way windows did in the past: convert the input parameter (the search path) to a wide string, call FindFirstFileExW_rep with that parameter and a temporary output variable, then transfer the values from the temporary output variable to the "real" output, again converting string parameters back to ansi. This way the ...ExW hook still does all the rerouting.

Ideally you should also implement FindFirstFileW and FindFirstFileExA hooks in a similar manner, although the latter would probably be a bit more complicated because it can have three different types of output formats, so you'd need three code paths for the conversion of the output parameter.

**Comment by ThosRTanner** [comment]
Could this affect the FindNextFile call as well?

**Comment by TanninOne** [comment]
Possibly, you'd have to test. But once you have the code to convert WIN32_FIND_DATAW to WIN32_FIND_DATAA, implementing FindNextFileA shouldn't be a lot of trouble.

**Comment by Tyler799** [comment]
Oh man, I hope someone fixes this. No way in heck am I going to use NMM again! If there was a fix, would it be published separately on its own nexus page, or would the current one (https://www.nexusmods.com/skyrim/mods/1334/) be updated?

**Comment by chrisbro-MSFT** [comment]
Anyone already have an MO1 build environment set up?  I've started to piece together a potential fix but can't install the dependencies on my current dev box so I'll need a new VM to build myself.

**Comment by chrisbro-MSFT** [comment]
I believe the affected APIs are FindFirstFile*, GetFileAttributes*, SetFileAttributes*, CreateDirectory*, and DeleteFile*.  It is possible that MoveFile* is also affected; I can't tell from just reading the code.  FindNextFile looks like it should be ok as-is.

**Comment by Alexei-Ilio** [comment]
Just a heads up, I'm in the slow schedule for Windows Insider and it just updated to build 16278. Unfortunately, MO1 still doesn't work.

**Comment by ThosRTanner** [comment]
@chrisbro-MSFT I do though I'm not on the windows update schedule.

**Comment by Hacker432** [comment]
@chrisbro-MSFT
@ThosRTanner
I'm on 16278
I can try to setup the build environment

**Comment by ZeroGravity83** [comment]
Sorry to ask a question here but is it possible to go back from Build 16232 to Creators Update or Build 16232 without losing data? I did some googling but couldn't find a way to do so. I really regret joining the Windows Insider Program. I wanted to stay on Build 16232 for the 4GB VRAM fix, but it eventually forced me to update to the latest build and now I can't use MO anymore.

**Comment by Hacker432** [comment]
Tonight Windows has been updated again, now I am on build 16281
I have downloaded everything but I can't compile, I'm having this error
http://hastebin.com/nadepivuho.tex

**Comment by GrantSP** [comment]
I have several users in the STEP forum that will be able to test any changed code that someone can provide.
MO/MO2 users are more and more posting that it is simply not functioning anymore.

**Comment by chrisbro-MSFT** [comment]
Sorry, I didn't have much time to work on it while on vacation last week.  Fortunately, the Windows team has accepted a fix and it went into a build today.  I think that build 16289+ should have the fix.  (I didn't work on it myself, I'm just a silly Office dev, I'm just passing along the good news)

If it turns out that it isn't fixed, though, then please let us know ASAP!

**Comment by TucoPacifico1980** [comment]
@chrisbro-MSFT Oh that is fantastic news. I am merely following this issue having no coding skills. Just want to thank all actively involved for what you are doimg for us users.

**Comment by jenatali** [comment]
Looks like this fix is now available in the latest insider build. https://blogs.windows.com/windowsexperience/2017/09/12/announcing-windows-10-insider-preview-build-16288-pc-build-15250-mobile/

> We fixed an issue resulting in Mod Organizer no longer load mods into Skyrim in recent flights.

**Comment by feigi** [comment]
This would be so awesome. I hope it will work and my highly modded Skyrim finally run's smoothly. Thanks to all people involved! Crossing fingers.

**Comment by Tolrem** [comment]
I've just launched Skyrim with MO on build 16288 and can confirm it's working well :)

**Comment by Alexei-Ilio** [comment]
sweet, can't wait to try it when i get home

**Comment by atm0915** [comment]
I believe this issue can be marked solved?

**Comment by feigi** [comment]
I can confirm build 16288 works.

**Comment by TucoPacifico1980** [comment]
MS showed us love, time to switch on the Insider. Thanks tyo all of you. Downloading 16288 and installing 4K textures in New Vegas 💃 
Another Microsoft Windows customer:
https://imgur.com/a/DW6uG

**Comment by darkconsole** [comment]
confirming that 16288 makes all the things better.

**Comment by TucoPacifico1980** [comment]
Have you guys tried running Wrye Bash 307.beta1 available here:
https://www.nexusmods.com/skyrimspecialedition/mods/6837/?

... From Mod Organizer 2?

I get this:
https://imgur.com/a/C2S0k

Running Windows 10 build 16288. On default windows (non-Insider) it was working fine.  tried reinstalling Wrye Bash already.

**Comment by TucoPacifico1980** [comment]
I informed MS about the Wrye Bash problem: 
https://aka.ms/Ov5soz

Can someone check if you can start Wrye Bash from Mod Organizer 2 in Win10 16288?

**Comment by TucoPacifico1980** [comment]
For those with Windows Insider:

Emergency Video for Skyrim SE, Mod Organizer 2 and Wrye Bash in Windows 10 Insider build16288

https://www.reddit.com/r/skyrimmods/comments/7123jm/emergency_video_for_skyrim_se_mod_organizer_2_and/

**Comment by TucoPacifico1980** [comment]
@TanninOne is there an alternative place I cwn get Mod Organizer 2, as you've taken it off the Nexus?

**Comment by Dwagl** [comment]
Problem is back with the released Windows 10 Fall Update  (v1709, Build 16299.15).

SkyrimSE through MO2 won't load mods, configs or game controllers.

**Comment by darkconsole** [comment]
@Dwagl i been using this fork. https://github.com/LePresidente/modorganizer/releases

its super screwed up, spews errors basically you have to relaunch mo2 after each app you run through it. but the game still works fine through it for me.

**Comment by Ashes2Asherz** [comment]
Just wanted to let you guys know, I'm not experiencing any of these issues mentioned, everything is running very smooth for me.. I use MO 1.3.11

**Comment by darkconsole** [comment]
not for special edition you don't. mo1 cannot bootstrap 64bit binaries.

**Comment by Ashes2Asherz** [comment]
oh no I don't use sse.. just wanted to put out there I don't have any problems with 1.3.11 on LE with the fall update

**Comment by darkconsole** [comment]
yeah sorry, realized milliseconds after posting this was not the SSE bug report.
